vp8_remb_simulcast.c

```diff
@@ -0,0 +1,571 @@
+ * - IOS Safari (14.6) - No REMB ramp up when sending H 30fps - proper remb value right from the start
+ *
+ * Whats missing (and currently not target with the PR):
+ * - Expose the currently seen REMB value in the general peerConnection statistics (just missing for completeness as other pc related values are exposed and helpful for external debugging)
+ * - REMB probing on the subscription peerconnection (would speed up the REMB detection and would allow a switching without changing the temporal layers)
+ * - transport-cc which would dramatically improve the probing (hard to implement especially in C, so nothing we target now)
+ * - Gather the currently really used bitrates directly from the publishers peerconnection and not from the configure/join values
+ * - Support for VP9 (the current approach only targets VP8)
+ * - We currently start to send the layers as requested by the client
+ *
+ * - if i configure or start a subscription i want o be able to:
+ * - - force a certain higher layer to start with (if i am using chromium based or safari, that works out of the box)
+ * - - tell janus to achive that layer through remb ramping up
+ * - - tell janus not to do it for this subscription (e.g. i have ones that must send a certain layer)
+ */
+
+#include "vp8_remb_simulcast.h"
+#include "ice.h"
+
+/* Enables console debugging */
+#define REMB_CONSOLE_DEBUGGING 1
+/* This is a correction factor we use to calculate the bitrate if we transport 15fps instead of 3) *
+   The bitrates we handshake with the client are for 30fps, so we need to calculate the values for 15fps internally */
+#define CORRECTION_FACTOR_15FPS 0.6
+/* If we ramp up and need to ramp down within that period of time we call this ramp up failed */
+#define TIME_RAMP_UP_FAILED 7
+/* Defines how much we increase the bitrate for the failing layer per fail (10%) */
+#define REQUIRED_BITRATE_INCREASE_PER_FAIL 0.1
+
+/* Forward declarations */
+uint32_t janus_vp8_remb_simulcast_get_client_requested_ramp_position(janus_rtp_simulcasting_context *pSim);
+uint32_t janus_vp8_remb_simulcast_get_current_ramp_position(janus_rtp_simulcasting_context *pSim);
+uint32_t janus_vp8_remb_simulcast_get_bitrate_for_ramp_position(janus_rtp_simulcasting_remb_context *pRembContext, uint32_t rampPos);
+void janus_vp8_remb_simulcast_get_neighbour_ramp_positions(janus_rtp_simulcasting_context *pSim, uint32_t rampPos, uint32_t *pNextLower, uint32_t *pNextUpper);
+double janus_vp8_remb_simulcast_calculate_bitrate_change(janus_vp8_remb_subscriber *pSubscriber, uint32_t remb);
+void janus_vp8_remb_simulcast_update_last_bitrate_change(janus_rtp_simulcasting_remb_context *pRempContext, double change);
+gboolean janus_vp8_remb_simulcast_set_flags_for_ramp_position(janus_rtp_simulcasting_remb_context *pRembContext, uint32_t position);
+const char* janus_vp8_remb_simulcast_get_debug_substream_to_text(uint32_t substream);
+const char* janus_vp8_remb_simulcast_get_debug_temporal_to_text(uint32_t temporal);
```

Comment:
Code style: I've noticed this in a few other areas of the code as well, our code style for pointers is

    type *pointer

not

    type* pointer

and the same for functions that return a pointer, so

    type *function(..)